From 39b96f2d71edffc330956ac3875bf2551a7ff816 Mon Sep 17 00:00:00 2001
From: Ruchi Kandoi <kandoiruchi@google.com>
Date: Mon, 7 Jan 2019 16:17:48 -0800
Subject: [PATCH 04/10] Fix heap overflow in NFA_SendRawFrame()

Bug: 120664978
Test: NFC enable/disable
Change-Id: I1b6a062fb5bf10364a20e99faf4adef13a478d22
(cherry picked from commit e873f932bab2335edb2503dd775fa51d3ae99f01)
---
 src/nfa/dm/nfa_dm_api.c | 7 +++++++
 1 file changed, 7 insertions(+)

diff --git a/src/nfa/dm/nfa_dm_api.c b/src/nfa/dm/nfa_dm_api.c
index d2751ad..6e4a4df 100644
--- a/src/nfa/dm/nfa_dm_api.c
+++ b/src/nfa/dm/nfa_dm_api.c
@@ -28,6 +28,7 @@
 #include "nfa_dm_int.h"
 #include "nfa_ce_int.h"
 #include "nfa_sys_int.h"
+#include <log/log.h>
 #include "ndef_utils.h"
 
 /*****************************************************************************
@@ -869,6 +870,12 @@ tNFA_STATUS NFA_SendRawFrame (UINT8  *p_raw_data,
         return (NFA_STATUS_INVALID_PARAM);
 
     size = BT_HDR_SIZE + NCI_MSG_OFFSET_SIZE + NCI_DATA_HDR_SIZE + data_len;
+    /* Check for integer overflow */
+    if (size < data_len) 
+    {
+        android_errorWriteLog(0x534e4554, "120664978");
+        return NFA_STATUS_INVALID_PARAM;
+    }
     if ((p_msg = (BT_HDR *) GKI_getbuf (size)) != NULL)
     {
         p_msg->event  = NFA_DM_API_RAW_FRAME_EVT;
-- 
2.7.4

